### PR TITLE
[Crypto] Pairing edge case

### PR DESCRIPTION
**Pull request by tarakby**
- Relic multi-pairing computation has a bug in a specific edge case, that can be triggered by BLS private keys equal to 1 or -1 during signature verification. This PR introduces a temporary fix in Flow crypto while Relic gets fixed.
- add tests in the crypto lib to check the edge case.

### CODE REVIEW

File: crypto/bls_test.go [hunk 1]
```diff
@@ -51,6 +51,34 @@ func TestBLSMainMethods(t *testing.T) {
...
+			input := make([]byte, 100)
+			_, err = mrand.Read(input)
+			require.NoError(t, err)
+			s, err := sk.Sign(input, hasher)
+			require.NoError(t, err)
+			pk := sk.PublicKey()
+
+			// test a valid signature
+			result, err := pk.Verify(s, input, hasher)
+			assert.NoError(t, err)
```
Comment (durkmurder) [inline crypto/bls_test.go:77]:
```suggestion
			require.NoError(t, err)
```

File: crypto/sign_test_utils.go [hunk 2]
```diff
@@ -153,6 +153,10 @@ func testKeyGenSeed(t *testing.T, salg SigningAlgorithm, minLen int, maxLen int)
 	})
 }
 
+var BLS12381Order = []byte{0x73, 0xED, 0xA7, 0x53, 0x29, 0x9D, 0x7D, 0x48, 0x33, 0x39,
```
Comment (durkmurder) [inline crypto/sign_test_utils.go:156]:
```suggestion
const BLS12381Order = []byte{0x73, 0xED, 0xA7, 0x53, 0x29, 0x9D, 0x7D, 0x48, 0x33, 0x39,
```

**Review by durkmurder — APPROVED** [review]
LGTM

File: crypto/bls_test.go (hunk 1 above)
Comment (tarakby) [reply crypto/bls_test.go:77]: Well, the purpose of the test is to make sure the function does not error in this case. Before this PR's fix, the function errors in that case while it shouldn't.

File: crypto/sign_test_utils.go (hunk 2 above)
Comment (tarakby) [reply crypto/sign_test_utils.go:156]: I don't think it's possible to make declare an array as a constant, an array in Go cannot be immutable.

File: crypto/bls_test.go (hunk 1 above)
Comment (jordanschalm) [reply crypto/bls_test.go:77]: Whether to use `assert` or `require` depends on whether it is useful/meaningful to also check subsequent assertions, assuming this one failed. In this case it probably isn't useful (typically an error return means that value return `result` isn't meaningful). In other words, this test output:
```
Expected no error ...
Expected result to be true
```

is not more useful than just
```
Expected no error ...
```

In this case it doesn't make much difference IMO because there is only one subsequent assertion.

**Review by jordanschalm — APPROVED** [review]

File: crypto/bls_test.go [hunk 3]
```diff
@@ -51,6 +51,34 @@ func TestBLSMainMethods(t *testing.T) {
...
+			assert.True(t, result, fmt.Sprintf(
+				"Verification should succeed:\n signature:%s\n message:%x\n private key:%s", s, input, sk))
```
[the remark this answers is not in the document]
Comment (tarakby) [reply crypto/bls_test.go:79]: right, I think this also works:
```
assert.True(t, result,
				"Verification should succeed:\n signature:%s\n message:%x\n private key:%s", s, input, sk))
```

**Comment by tarakby** [comment]
bors merge

**Comment by tarakby** [comment]
bors merge
